### PR TITLE
Extract Gateway Code From Kubo

### PR DESCRIPTION
**Pull request by hacdias**
_The gateway has arrived 🚀._

This extracts the gateway from Kubo, as per tracked in https://github.com/ipfs/kubo/issues/8524. This PR is based on the history and code from https://github.com/ipfs/kubo/pull/9588.

### CODE REVIEW

**Comment by hacdias** [comment]
@lidel this PR is ready for review. I don't understand why `gorelease` is giving that output, I run it locally and got the same. However, all tests pass and importing this package from https://github.com/ipfs/kubo/pull/9588 works and everything is working properly. I will write an update if I find what the problem is.

**Comment by hacdias** [comment]
**UPDATE**: I have removed the version bump from this PR. I think it's best to bump the version after this PR is merged instead, so I can still easily work on possible improvements, such as the CI sharness tests I am adding now.

@lidel this PR is ready for review. I don't understand why gorelease is giving that output, I run it locally and got the same. However, all tests pass and importing this package from https://github.com/ipfs/kubo/pull/9588 works and everything is working properly. I will write an update if I find what the problem is (go mod tidy and go get -d ./... all return status 0 and no changes).

As far as I understand, per the gorelease command above, it is downloading go-libipfs v0.3.0 and then trying to import the new gateway module. And it of course fails because the new gateway module is... new. AFAIK, it should instead say that there is a new module called gateway but I can't understand why it doesn't.

I even manually did what `gorelease` does:

1. Create a temporary package.
2. Import every subpackage of `go-libipfs`
3. Use `replace` in `go.mod` to point to the this new version.
4. Run `go get -d .`

And it gave no errors. So I don't really know what's going on here.

File: .github/workflows/gateway-sharness.yml
Comment (hacdias) [inline .github/workflows/gateway-sharness.yml:33]: MUST FIX: Replace go.mod to use this PRs go-libipfs.

**Comment by hacdias** [comment]
PS: I'm still working on the CI. https://github.com/ipfs/go-libipfs/pull/138

File: .github/workflows/gateway-sharness.yml
Comment (lidel) [inline .github/workflows/gateway-sharness.yml:41]: :thought_balloon: this will probably break when we work on https://github.com/ipfs/go-libipfs/issues/61 but after we stabilize GO API, it will be ok

File: gateway/README.md
Comment (lidel) [inline gateway/README.md]:
```suggestion
> A reference implementation of HTTP Gateway Specifications.
```

File: gateway/README.md
Comment (lidel) [inline gateway/README.md]:
```suggestion
* Gateway Specification: https://github.com/ipfs/specs/tree/main/http-gateways#readme
* Types of HTTP Gateways: https://docs.ipfs.tech/how-to/address-ipfs-on-web/#http-gateways
```

File: gateway/README.md
Comment (lidel) [inline gateway/README.md]:
```suggestion
// Create http mux and setup path gateway handler.
```

File: gateway/README.md
Comment (lidel) [inline gateway/README.md]:
```suggestion
// Start the server on :8080 and voilá! You have a basic IPFS gateway running
```

File: gateway/README.md
Comment (lidel) [inline gateway/README.md:1]:
```suggestion
# HTTP Gateway
```

**Review by lidel — APPROVED** [review]
Thank you @hacdias, especially for cleaning up assets and squeezing sharness in to give us some protection from the start.

Let's merge this to unblock https://github.com/ipfs/go-libipfs/issues/61 and follow-up work for `dweb.link` gateway.

**Pull request merged**
